### PR TITLE
PR: Add a way to handle font errors in QtAwesome

### LINKED ISSUE #3843: Spyder shows blank screen because of errors in QtAwesome
Running Windows 10 on a 64 Bit machine. 

**Steps:**
1. İnstall Anaconda (Anaconda3-4.2.0-Windows-x86_64.exe)
2. Launch Spyder 3.0.0 within Anaconda Navigator

_Note: At the very beginning -first time- spyder started succesfully. i could even install python modules and compiled a piece of code. After first restart of machine, problems started to occur as i mentioned below_

**Results:**
- Spyder crashes with lots of qt log in console.

- Spyder icon in Anaconda navigator also crashes.

`spyder --show-console`
```
Font is empty
Font is empty
Font is empty
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QWindowsWindow::setGeometry: Unable to set geometry 1260x1442+18+41 on QWidgetWindow/'MainWindowClassWindow'. Resulting geometry:  1260x1061+18+41 (frame: 8, 31, 8, 8, custom margin: 0, 0, 0, 0, minimum size: 418x716, maximum size: 16777215x16777215).
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
QPainter::end: Painter ended with 2 saved states
```

It looks like a qt related thing. I searched existing issues and already tried:
`conda update qt pyqt` 
```
pyqt                      5.6.0                    py35_1
qt                        5.6.2                    vc14_0  [vc14]
```

`python` 
Python 3.5.2 |Anaconda 4.2.0 (64-bit)| (default, Jul  5 2016, 11:41:13) [MSC v.1900 64 bit (AMD64)] on win32

`ipython qtconsole` (works fine)
`python spyder --reset` 
`python spyder --new_instance` (same crash)

At last, i updated spyder to 3.0.2 but nothing changes.

I installed Anaconda (after uninstalling carefully -with anaconda clean-) at least 5 times. https://docs.continuum.io/anaconda/install

_Last edit before any comment_
_I'm able to run spyder (via anaconda) succefully on another windows 7 pc (64 bit) for a long time. The difference is: i installed theano and keras libraries on this mentioned machine after first installation and able to run codes from these libs succesfully till first shut down of machine._

Tried below solution and get spyder work at one time but without any icon in IDE:
`conda install qt=4.8.7=vc10_4` ( as mentioned here https://github.com/spyder-ide/spyder/issues/2858)

## Dependencies

```
#
_license                  1.1                      py35_1
_nb_ext_conf              0.3.0                    py35_0
alabaster                 0.7.9                    py35_0
anaconda                  4.2.0               np111py35_0
anaconda-clean            1.0.0                    py35_0
anaconda-client           1.5.1                    py35_0
anaconda-navigator        1.3.1                    py35_0
argcomplete               1.0.0                    py35_1
astroid                   1.4.7                    py35_0
astropy                   1.2.1               np111py35_0
babel                     2.3.4                    py35_0
backports                 1.0                      py35_0
beautifulsoup4            4.5.1                    py35_0
bitarray                  0.8.1                    py35_1
blaze                     0.10.1                   py35_0
bokeh                     0.12.2                   py35_0
boto                      2.42.0                   py35_0
bottleneck                1.1.0               np111py35_0
bzip2                     1.0.6                    vc14_3  [vc14]
certifi                   2016.2.28                py35_0
cffi                      1.7.0                    py35_0
chest                     0.2.3                    py35_0
click                     6.6                      py35_0
cloudpickle               0.2.1                    py35_0
clyent                    1.2.2                    py35_0
colorama                  0.3.7                    py35_0
comtypes                  1.1.2                    py35_0
conda                     4.2.13                   py35_0
conda-build               2.0.2                    py35_0
conda-env                 2.6.0                         0
configobj                 5.0.6                    py35_0
console_shortcut          0.1.1                    py35_1
contextlib2               0.5.3                    py35_0
cryptography              1.6                      py35_0
curl                      7.49.0                   vc14_0  [vc14]
cycler                    0.10.0                   py35_0
cython                    0.24.1                   py35_0
cytoolz                   0.8.0                    py35_0
dask                      0.11.0                   py35_0
datashape                 0.5.2                    py35_0
decorator                 4.0.10                   py35_0
dill                      0.2.5                    py35_0
docutils                  0.12                     py35_2
dynd-python               0.7.2                    py35_0
entrypoints               0.2.2                    py35_0
et_xmlfile                1.0.1                    py35_0
fastcache                 1.0.2                    py35_1
filelock                  2.0.6                    py35_0
flask                     0.11.1                   py35_0
flask-cors                2.1.2                    py35_0
freetype                  2.5.5                    vc14_1  [vc14]
get_terminal_size         1.0.0                    py35_0
gevent                    1.1.2                    py35_0
greenlet                  0.4.10                   py35_0
h5py                      2.6.0               np111py35_2
hdf5                      1.8.15.1                 vc14_4  [vc14]
heapdict                  1.0.0                    py35_1
icu                       57.1                     vc14_0  [vc14]
idna                      2.1                      py35_0
imagesize                 0.7.1                    py35_0
ipykernel                 4.5.0                    py35_0
ipython                   5.1.0                    py35_0
ipython_genutils          0.1.0                    py35_0
ipywidgets                5.2.2                    py35_0
itsdangerous              0.24                     py35_0
jdcal                     1.2                      py35_1
jedi                      0.9.0                    py35_1
jinja2                    2.8                      py35_1
jpeg                      8d                       vc14_2  [vc14]
jsonschema                2.5.1                    py35_0
jupyter                   1.0.0                    py35_3
jupyter_client            4.4.0                    py35_0
jupyter_console           5.0.0                    py35_0
jupyter_core              4.2.0                    py35_0
lazy-object-proxy         1.2.1                    py35_0
libdynd                   0.7.2                         0
libpng                    1.6.22                   vc14_0  [vc14]
libtiff                   4.0.6                    vc14_2  [vc14]
llvmlite                  0.13.0                   py35_0
locket                    0.2.0                    py35_1
lxml                      3.6.4                    py35_0
markupsafe                0.23                     py35_2
matplotlib                1.5.3               np111py35_0
menuinst                  1.4.1                    py35_0
mistune                   0.7.3                    py35_0
mkl                       11.3.3                        1
mkl-service               1.1.2                    py35_2
mpmath                    0.19                     py35_1
multipledispatch          0.4.8                    py35_0
nb_anacondacloud          1.2.0                    py35_0
nb_conda                  2.0.0                    py35_0
nb_conda_kernels          2.0.0                    py35_0
nbconvert                 4.2.0                    py35_0
nbformat                  4.1.0                    py35_0
nbpresent                 3.0.2                    py35_0
networkx                  1.11                     py35_0
nltk                      3.2.1                    py35_0
nose                      1.3.7                    py35_1
notebook                  4.2.3                    py35_0
numba                     0.28.1              np111py35_0
numexpr                   2.6.1               np111py35_0
numpy                     1.11.1                   py35_1
odo                       0.5.0                    py35_1
openpyxl                  2.3.2                    py35_0
openssl                   1.0.2j                   vc14_0  [vc14]
pandas                    0.18.1              np111py35_0
partd                     0.3.6                    py35_0
path.py                   8.2.1                    py35_0
pathlib2                  2.1.0                    py35_0
patsy                     0.4.1                    py35_0
pep8                      1.7.0                    py35_0
pickleshare               0.7.4                    py35_0
pillow                    3.3.1                    py35_0
pip                       8.1.2                    py35_0
pkginfo                   1.3.2                    py35_0
ply                       3.9                      py35_0
prompt_toolkit            1.0.3                    py35_0
psutil                    4.3.1                    py35_0
py                        1.4.31                   py35_0
pyasn1                    0.1.9                    py35_0
pycosat                   0.6.1                    py35_1
pycparser                 2.14                     py35_1
pycrypto                  2.6.1                    py35_4
pycurl                    7.43.0                   py35_0
pyflakes                  1.3.0                    py35_0
pygments                  2.1.3                    py35_0
pylint                    1.5.4                    py35_1
pyopenssl                 16.0.0                   py35_0
pyparsing                 2.1.4                    py35_0
pyqt                      5.6.0                    py35_1
pytables                  3.2.2               np111py35_4
pytest                    2.9.2                    py35_0
python                    3.5.2                         0
python-dateutil           2.5.3                    py35_0
pytz                      2016.6.1                 py35_0
pywin32                   220                      py35_1
pyyaml                    3.12                     py35_0
pyzmq                     15.4.0                   py35_0
qt                        5.6.2                    vc14_0  [vc14]
qtawesome                 0.3.3                    py35_0
qtconsole                 4.2.1                    py35_2
qtpy                      1.1.2                    py35_0
requests                  2.11.1                   py35_0
rope                      0.9.4                    py35_1
ruamel_yaml               0.11.14                  py35_0
scikit-image              0.12.3              np111py35_1
scikit-learn              0.17.1              np111py35_1
scipy                     0.18.1              np111py35_0
setuptools                27.2.0                   py35_1
simplegeneric             0.8.1                    py35_1
singledispatch            3.4.0.3                  py35_0
sip                       4.18                     py35_0
six                       1.10.0                   py35_0
snowballstemmer           1.2.1                    py35_0
sockjs-tornado            1.0.3                    py35_0
sphinx                    1.4.6                    py35_0
spyder                    3.0.2                    py35_0
sqlalchemy                1.0.13                   py35_0
statsmodels               0.6.1               np111py35_1
sympy                     1.0                      py35_0
tk                        8.5.18                   vc14_0  [vc14]
toolz                     0.8.0                    py35_0
tornado                   4.4.1                    py35_0
traitlets                 4.3.0                    py35_0
unicodecsv                0.14.1                   py35_0
vs2015_runtime            14.0.25123                    0
wcwidth                   0.1.7                    py35_0
werkzeug                  0.11.11                  py35_0
wheel                     0.29.0                   py35_0
widgetsnbextension        1.2.6                    py35_0
win_unicode_console       0.5                      py35_0
wrapt                     1.10.6                   py35_0
xlrd                      1.0.0                    py35_0
xlsxwriter                0.9.3                    py35_0
xlwings                   0.10.0                   py35_0
xlwt                      1.1.2                    py35_0
zlib                      1.2.8                    vc14_3  [vc14]
```

### PR DESCRIPTION
**Pull request by dalthviz**
Fixes #3843 

----
Is important to know that for this to actually work this [PR](https://github.com/spyder-ide/qtawesome/pull/60) has to get a merge in the qtawesome project because of the `FontError` import.

### CODE REVIEW

**Comment by dalthviz** [comment]
A preview of the message dialog:

corruptfont

File: spyder/app/mainwindow.py [hunk 1]
```diff
@@ -2966,6 +2968,24 @@ def main():
     mainwindow = None
     try:
         mainwindow = run_spyder(app, options, args)
+    except FontError as fontError:
+        import traceback
+        traceback.print_exc(file=STDERR)
+        traceback.print_exc(file=open('spyder_crash.log', 'w'))
```
Comment (ccordoba12) [inline spyder/app/mainwindow.py]: We don't need these three lines.

File: spyder/app/mainwindow.py [hunk 2]
```diff
@@ -2966,6 +2968,24 @@ def main():
...
+                "bug report</u>, please try to replace the fonts files in the "
+                "route above with the .ttf files from <a href=\"%s\">here</a>."
+                "<br><br>"
+                "If trying this does not help, please take "
+                "the time to search for <a href=\"%s\">known bugs</a> or "
+                "<a href=\"%s\">discussions</a> matching your situation before"
+                " eventually creating a new issue <a href=\"%s\">here</a>. "
+                "Your feedback will always be greatly appreciated."
+                "" % (fontError ,__project_fonts__, __project_url__,
+                      __forum_url__, __project_url__))
```
Comment (ccordoba12) [inline spyder/app/mainwindow.py]: Please rephrase this text like this

```
Spyder can't load the <i>Spyder 3</i> icon theme. That's why it's going to
fallback to the theme used in Spyder 2.

For that, please close this window and start Spyder again.
```

and make it translatable.

Comment (ccordoba12) [inline spyder/app/mainwindow.py]: And after this dialog, we need to add this line

    CONF.set('main', 'icon_theme', 'spyder 2')

(but please test that my approach works correctly by removing the QtAwesome fonts in your computer and starting Spyder again).

Comment (ccordoba12) [inline spyder/app/mainwindow.py]: Better message

Spyder was unable to load the <i>Spyder 3</i> icon theme. That's why it's going to
fallback to the theme used in Spyder 2.

For that, please close this window and start Spyder again.
```

**Comment by ccordoba12** [comment]
@dalthviz, I added a commit to your branch to fix some Travis errors. Please don't forget to pick it up locally.

Still this is a bit tricky to merge because it depends on a new version of QtAwesome. I'm going to release QtAwesome **0.4** with your `FontError` fix, so @dalthviz you need to bump here too our package requirements to that version in all places we define them. And for that, please check how @rlaverde did it for Chardet in PR #3640.

**Comment by goanpeca** [comment]
@ccordoba12 regarding

>...you need to bump here too our package requirements to that version in all places we define them.

I suggest we use a requirements.txt file (or more as needed) as a single entry point for modifying dependencies (we can rely on this file for all CI services and scripts). This will make the updating of dependencies less error prone. I implemented this for fusion and eases the pain :0)

**Comment by ccordoba12** [comment]
> I suggest we use a requirements.txt file

Three reasons i think this is not a good idea:

1. We need fine grained control for our dependencies. For example, in CircleCI we need to install qtconsole along with the right PyQt version separately from the rest of our deps.
2. We're still testing that our conda and pip packages pull the right dependencies on Travis and AppVeyor. And for that, we don't need a requirements file, just let the packages to pull the dependencies.
3. People still wants to know (in README and our docs) what our dependencies are. Again, we can't use a simple requirements file for that.

I know this process is still cumbersome, but that's why CI services are for: to test if we did things right :-)

**Comment by goanpeca** [comment]
Regarding the reasons

1.) That is the exception and not the rule
2.) So?
3.) Sure we can, we put a link to the file and comment out the dependencies we have to handle "manually" on CI, which is basically for now just PyQt

**Comment by goanpeca** [comment]
>... reasons i think this is not a good idea:

If I had nickel for every time...

**Comment by ccordoba12** [comment]
@dalthviz, please add the QtAwesome version we need now in `README.md` and `docs/installation.rst` as **0.4+**.

@goanpeca, sometimes you're right and sometimes not. But you're prone to give me the task of finishing what you start, and I really (really) don't have time for that now. So I think we're going to continue with our awkward (but proven ;-) procedure. As I said before, it's not something we do everyday so I'm fine with it.

**Comment by goanpeca** [comment]
image

**Comment by ccordoba12** [comment]
@dalthviz, please change QtAwesome version to **0.4.1**. I had to do some changes in 0.4 to better detect this error.

File: spyder/app/mainwindow.py [hunk 3]
```diff
@@ -145,6 +145,7 @@
                               PY3, qbytearray_to_str, configparser as cp)
 from spyder.utils import encoding, programs
 from spyder.utils import icon_manager as ima
+from qtawesome.iconic_font import FontError
```
Comment (ccordoba12) [inline spyder/app/mainwindow.py]: Please move this import after the qtpy ones, i.e. after line **92**

**Comment by ccordoba12** [comment]
Ok, this is ready now. Thanks a lot @dalthviz for your work on this one. It's very important for 3.1 :-)

**Pull request merged**